internal/resource/url.go

```diff
@@ -319,6 +332,44 @@ func (f *Fetcher) fetchFromDataURL(u url.URL, dest io.Writer, opts FetchOptions)
 	return f.decompressCopyHashAndVerify(dest, bytes.NewBuffer(url.Data), opts)
 }
 
+// FetchFromGCS writes the data stored in a GCS bucket as described by u into dest,returning
+// an error if one is encountered. It looks for the default credentials by querying metadata
+// server on GCE. If it fails to get the credentials, then it will fall back to anonymous
+// credentials to fetch the object content.
+func (f *Fetcher) fetchFromGCS(u url.URL, dest io.Writer, opts FetchOptions) error {
+	ctx := context.Background()
+	var clientOption option.ClientOption
+	if f.GCSSession == nil {
+		if metadata.OnGCE() {
+			id, _ := metadata.ProjectID()
+			creds := &google.Credentials{
+				ProjectID:   id,
+				TokenSource: google.ComputeTokenSource("", storage.ScopeReadOnly),
+			}
+			clientOption = option.WithCredentials(creds)
+		} else {
+			f.Logger.Debug("falling back to unauthenticated GCS access")
+			clientOption = option.WithoutAuthentication()
+		}
+
+		var err error
+		f.GCSSession, err = storage.NewClient(ctx, clientOption)
+		if err != nil {
+			return err
+		}
+	}
+
+	path := strings.TrimLeft(u.Path, "/")
+	ctx, cancel := context.WithTimeout(ctx, time.Second*50)
+	defer cancel()
+	rc, err := f.GCSSession.Bucket(u.Host).Object(path).NewReader(ctx)
+	if err != nil {
+		return fmt.Errorf("couldn't find the bucket (%q): %v", path, err)
```

Comment:
Is this really a problem finding the bucket?

We should probably give the full URL here, not just the path.

Reply:
It is one of the problems that we may encounter.